python/zfp.pyx

```diff
@@ -0,0 +1,423 @@
+            compress_str = (<char *>data)[:compressed_size]
+    finally:
+        zfp_field_free(field)
+        zfp_stream_close(stream)
+        stream_close(bstream)
+
+    return compress_str
+
+cdef view.array _decompress_with_view(
+    zfp_field* field,
+    zfp_stream* stream,
+):
+    cdef zfp_type ztype = field[0]._type
+    dtype = ztype_to_dtype(ztype)
+    format_type = dtype_to_format(dtype)
+
+    shape = (field[0].nw, field[0].nz, field[0].ny, field[0].nx)
+    shape = tuple([x for x in shape if x > 0])
+
+    cdef view.array decomp_arr = view.array(
+        shape,
+        itemsize=np.dtype(dtype).itemsize,
+        format=format_type,
+        allocate_buffer=True
+    )
+    cdef void* pointer = <void *> decomp_arr.data
+    with nogil:
+        zfp_field_set_pointer(field, pointer)
+        ret = zfp_decompress(stream, field)
+    if ret == 0:
+        raise RuntimeError("error during zfp decompression")
+    return decomp_arr
+
+cdef _decompress_with_user_array(
+    zfp_field* field,
+    zfp_stream* stream,
+    void* out,
+):
+    with nogil:
+        zfp_field_set_pointer(field, out)
+        ret = zfp_decompress(stream, field)
+    if ret == 0:
+        raise RuntimeError("error during zfp decompression")
+
+cdef _set_compression_mode(
+    zfp_stream *stream,
+    zfp_type ztype,
+    int ndim,
+    double tolerance = -1,
+    double rate = -1,
+    int precision = -1,
+):
+    if tolerance >= 0:
+        zfp_stream_set_accuracy(stream, tolerance)
+    elif rate >= 0:
+        zfp_stream_set_rate(stream, rate, ztype, ndim, 0)
+    elif precision >= 0:
+        zfp_stream_set_precision(stream, precision)
+    else:
+        zfp_stream_set_reversible(stream)
+
+cdef _validate_4d_list(in_list, list_name):
+    # Validate that the input list is either a valid list for strides or shape
+    # Specifically, check it is a list and the length is > 0 and <= 4
+    # Throws a TypeError or ValueError if invalid
+    try:
+        if len(in_list) > 4:
+            raise ValueError(
+                "User-provided {} has too many dimensions "
+                "(up to 4 supported)"
+            )
+        elif len(in_list) <= 0:
+            raise ValueError(
+                    "User-provided {} needs at least one dimension"
+            )
+    except TypeError:
+        raise TypeError(
+            "User-provided {} is not an iterable"
+        )
+
+cdef _validate_userinput_matches_header(
+    zfp_field* field,
+    zfp_stream* stream,
+    out,
+    zfp_type ztype,
+    shape,
+    strides,
+    double tolerance,
+    double rate,
+    int precision
+):
+    # check that the header and user type matches
+    if ztype != type_none and ztype is not field[0]._type:
+        raise ValueError(
+            "User-provided zfp_type does not match zfp_type in header"
+        )
+
+    # check that the header and user shapes match
+    header_shape = (field[0].nx, field[0].ny, field[0].nz, field[0].nw)
```

Comment:
Usually we map nx to the rightmost dimension of a numpy array.  Should these be reversed?